schema-templates/basic-auth.sql

```diff
@@ -0,0 +1,298 @@
+$$;
+
+drop trigger if exists send_validation_t on basic_auth.logins;
+create trigger send_validation_t
+  after insert on basic_auth.logins
+  for each row
+  execute procedure basic_auth.send_validation();
+
+-------------------------------------------------------------------------------
+-- Email Validation and Password Reset
+
+create table if not exists
+basic_auth.tokens (
+  token       character varying unique,
+  token_type  varchar(64) not null,
+  username    character varying not null,
+  created_at  timestamptz not null default current_date,
+  constraint  t_pk primary key (token),
+  constraint  t_login_fk foreign key (username) references basic_auth.logins
+              on delete cascade on update cascade
+);
+
+-------------------------------------------------------------------------------
+-- Login helper
+
+create or replace function
+basic_auth.login_role(username text, pass text) returns text
+  language plpgsql
+  as $$
+begin
+  return (
+  select role from basic_auth.logins
+   where logins.username = login_role.username
+     and logins.pass = crypt(login_role.pass, logins.pass)
+  );
+end;
+$$;
+
+-------------------------------------------------------------------------------
+-- Public functions (in current schema, not basic_auth)
+
+create or replace function
+request_password_reset(username text) returns void
+  language plpgsql
+  as $$
+declare
+  tok character varying;
+begin
+  delete from basic_auth.tokens
+   where token_type = 'reset'
+     and tokens.username = request_password_reset.username;
+
+  select basic_auth.random_value(64) into tok;
+  insert into basic_auth.tokens (token, token_type, username)
+         values (tok, 'reset', request_password_reset.username);
+  perform pg_notify('reset',
+    json_build_object(
+      'email', (select email
+                  from basic_auth.logins
+                 where logins.username = request_password_reset.username),
+      'username', request_password_reset.username,
+      'token', tok,
+      'token_type', 'reset'
+    )::text
+  );
+end;
+$$;
+
+create or replace function
+reset_password(username text, token text, pass text)
+  returns void
+  language plpgsql
+  as $$
+declare
+  tok character varying;
+begin
+  if exists(select 1 from basic_auth.tokens
+             where tokens.username = reset_password.username
+               and tokens.token = reset_password.token
+               and token_type = 'reset') then
+    update basic_auth.logins set pass=reset_password.pass
+     where logins.username = reset_password.username;
+
+    delete from basic_auth.tokens
+     where tokens.username = reset_password.username
+       and tokens.token = reset_password.token
+       and token_type = 'reset';
+  else
+    raise invalid_password using message =
+      'invalid user or token';
+  end if;
+  delete from basic_auth.tokens
+   where token_type = 'reset'
+     and tokens.username = reset_password.username;
+
+  select basic_auth.random_value(64) into tok;
```

Comment:
Why are we creating a new token if reset fails? If that's what this code is doing, why not just reuse `request_reset_password`?